skbio/diversity/_base_cy.pyx

```diff
@@ -0,0 +1,215 @@
+
+    # apply the mask such that tip_ds only includes values which correspond to
+    # the tips of the tree.
+    for i in range(n_rows):
+        tip_ds[i] *= mask[i]
+
+    return tip_ds
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef _traverse_reduce(np.ndarray[DTYPE_t, ndim=2] child_index,
+                      np.ndarray[DTYPE_t, ndim=2] a):
+    """Apply a[k] = sum[i:j]
+
+    Parameters
+    ----------
+    child_index: np.array of int
+        A matrix in which the first column corresponds to an index position in
+        ``a``, which represents a node in a tree. The second column is the
+        starting index in ``a`` for the node's children, and the third column
+        is the ending index in ``a`` for the node's children.
+    a : np.ndarray of int
+        A matrix of the environment data. Each row corresponds to a node in a
+        tree, and each column corresponds to an environment. On input, it is
+        assumed that only tips have counts.
+
+    Notes
+    -----
+    This is effectively a postorder reduction over the tree. For example,
+    given the following tree:
+
+                            /-A
+                  /E-------|
+                 |          \-B
+        -root----|
+                 |          /-C
+                  \F-------|
+                            \-D
+
+    And assuming counts for [A, B, C, D] in environment FOO of [1, 1, 1, 0] and
+    counts for environment BAR of [0, 1, 1, 1], the input counts matrix ``a``
+    would be:
+
+        [1 0  -> A
+         1 1  -> B
+         1 1  -> C
+         0 1  -> D
+         0 0  -> E
+         0 0  -> F
+         0 0] -> root
+
+    The method will perform the following reduction:
+
+        [1 0     [1 0     [1 0     [1 0
+         1 1      1 1      1 1      1 1
+         1 1      1 1      1 1      1 1
+         0 1  ->  0 1  ->  0 1  ->  0 1
+         0 0      2 1      2 1      2 1
+         0 0      0 0      1 2      1 2
+         0 0]     0 0]     0 0]     3 3]
+
+    The index positions of the above are encoded in ``child_index`` which
+    describes the node to aggregate into, and the start and stop index
+    positions of the nodes immediate descendents.
+
+    This method operates inplace on ``a``
+    """
+    cdef:
+        Py_ssize_t i, j, k
+        DTYPE_t node, start, end
+        DTYPE_t n_envs = a.shape[1]
+
+    # possible GPGPU target
+    for i in range(child_index.shape[0]):
+        node = child_index[i, 0]
+        start = child_index[i, 1]
+        end = child_index[i, 2]
+
+        for j in range(start, end + 1):
+            for k in range(n_envs):
+                a[node, k] += a[j, k]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _nodes_by_counts(np.ndarray counts,
+                     np.ndarray tip_ids,
+                     dict indexed):
+    """Construct the count array, and the counts up the tree
+
+    Parameters
+    ----------
+    counts : np.array of int
```

Comment:
Do we definitely want to only accept integers as input?
It may be advantageous to allow for floats to be passed in.